File: ts/pila/src/pila-encadenada.ts
class Nodo {
  private _item: number;
  private _siguiente: Nodo | null;

  get item() {
    return this._item;
  }

  get siguiente() {
    return this._siguiente;
  }

  set item(item: number) {
    this._item = item;
  }

  set siguiente(siguiente: Nodo | null) {
    this._siguiente = siguiente;
  }
}

export class PilaEncadenada {
  private dimension: number;
  private cantidad: number;
  private tope: Nodo | null;

  constructor(dimension: number) {
    this.dimension = dimension;
    this.cantidad = 0;
    this.tope = null;
  }

  insertar(item: number): number {
    if (this.cantidad == this.dimension) {
      console.log("Pila llena.");
      return -1;
    }
    let nuevoNodo = new Nodo();
    nuevoNodo.item = item;
    nuevoNodo.siguiente = this.tope;
    this.tope = nuevoNodo;
    this.cantidad++;
    return this.tope.item;
  }

  suprimir(): number | null {
    if (this.vacia()) {
      console.log("Pila vacia.");
      return null;
    }
    let eliminado = this.tope;
    this.tope = this.tope!.siguiente;
    this.cantidad--;
    return eliminado!.item;
  }

  recorrer(): void {
    let actual = this.tope;
    let pila: number[] = [];
    while (actual != null) {
      pila.push(actual.item);
      actual = actual!.siguiente;
    }
    console.log(pila);
  }

  vacia(): boolean {
    return this.cantidad == 0;
  }
}

File: estructuras/PilaEncadenada.py
class Nodo:
  __item: int
  __siguiente: "Nodo"

  def __init__(self,item=0,siguiente=None):
    self.__item = item
    self.__siguiente = siguiente
  
  def get_item(self):
    return self.__item
  
  def get_siguiente(self):
    return self.__siguiente
  
  def set_item(self, item):
    self.__item = item

  def set_siguiente(self, siguiente):
    self.__siguiente = siguiente

class PilaEncadenada:
  __dimension:int
  __cantidad:int
  __tope:Nodo

  def __init__(self, dimension):
    self.__dimension = dimension 
    self.__cantidad = 0
    self.__tope = None

  def insertar(self, item):
    if self.__cantidad == self.__dimension:
      print("Pila llena.")
      return None
    nuevo_nodo = Nodo(item)
    nuevo_nodo.set_siguiente(self.__tope)
    self.__tope = nuevo_nodo
    self.__cantidad+=1
    return item
  
  def suprimir(self):
    if self.vacia():
      print("Pila vacia.")
      return None
    eliminado = self.__tope
    self.__tope = self.__tope.get_siguiente()
    self.__cantidad -=1
    return eliminado
  
  def recorrer(self):
    if self.vacia():
      print("Pila vacia.")
      return None
    actual = self.__tope
    while actual is not None:
      print(actual.get_item(), end=" ")
      actual = actual.get_siguiente()
    print("\n")
  
  def vacia(self):
    return self.__cantidad == 0

if __name__ == "__main__":
  pila = PilaEncadenada(4)

  pila.insertar(1)
  pila.insertar(2)
  pila.insertar(3)
  pila.insertar(4)
  pila.insertar(5)

  pila.recorrer()

  pila.suprimir()
  pila.suprimir()

  pila.recorrer()